src/metrics-collector/cf/cf_client_test.go

```diff
@@ -9,217 +10,245 @@ import (
+   "user": "38b2f682-04bf-48af-9e08-0325aa5c4ea9"
+}
+`)
+
+var authBody = []byte(`
+{
+	"access_token":"test-access-token",
+	"token_type":"bearer",
+	"refresh_token":"test-refresh-token",
+	"expires_in":12000,
+	"scope":"openid cloud_controller.read password.write cloud_controller.write",
+	"jti":"a735f90f-0b49-447d-8f9d-ae2fbc1491dd"
+}
+`)
+
 var _ = Describe("CfClient", func() {
 	Describe("Login", func() {
 		var (
-			fakeCC *ghttp.Server
-			cfc    CfClient
-			conf   *config.Config
-			err    error
+			fakeCC          *ghttp.Server
+			fakeLoginServer *ghttp.Server
+			cfc             CfClient
+			conf            *config.Config
+			err             error
 		)
 
 		BeforeEach(func() {
 			fakeCC = ghttp.NewServer()
-			conf = config.DefaultConfig()
+			fakeLoginServer = ghttp.NewServer()
+			conf = &config.Config{
+				Cf:      config.DefaultCfConfig,
+				Logging: config.DefaultLoggingConfig,
+				Server:  config.DefaultServerConfig,
+			}
+
 			conf.Cf.Api = fakeCC.URL()
 			err = nil
 		})
 
 		JustBeforeEach(func() {
-			cfc = NewCfClient(&conf.Cf)
+			cfc = NewCfClient(&conf.Cf, lager.NewLogger("cf"))
 			err = cfc.Login()
 		})
 
 		AfterEach(func() {
 			if fakeCC != nil {
 				fakeCC.Close()
 			}
+			if fakeLoginServer != nil {
+				fakeLoginServer.Close()
+			}
+
 		})
 
-		Context("when retrieving endpoints fail", func() {
+		Context("when retrieving endpoints fails", func() {
+			Context("when the request fails", func() {
+				BeforeEach(func() {
+					fakeCC.Close()
+					fakeCC = nil
+				})
+
+				It("should error", func() {
+					Expect(err).To(BeAssignableToTypeOf(&url.Error{}))
+					urlErr := err.(*url.Error)
+					Expect(urlErr.Err).To(BeAssignableToTypeOf(&net.OpError{}))
+				})
+			})
+
 			Context("when a non-200 status code is returned", func() {
 				BeforeEach(func() {
 					fakeCC.AppendHandlers(
 						ghttp.CombineHandlers(
-							ghttp.VerifyRequest("GET", "/v2/info"),
+							ghttp.VerifyRequest("GET", PATH_CF_INFO),
 							ghttp.RespondWith(500, ""),
 						),
 					)
 				})
 
-				It("returns an error", func() {
-					Expect(err).To(MatchError(MatchRegexp("Error retrieving cf endpoints: .*")))
+				It("should error", func() {
+					Expect(err).To(MatchError(MatchRegexp("Error requesting endpoints: .*")))
 				})
 			})
 
-			Context("when the request fails", func() {
+		})
+
+		Context("when retrieving endpoints succeeds", func() {
+			Context("when the auth url is not valid", func() {
 				BeforeEach(func() {
-					fakeCC.Close()
-					fakeCC = nil
+					fakeCC.AppendHandlers(
+						ghttp.CombineHandlers(
+							ghttp.VerifyRequest("GET", PATH_CF_INFO),
+							ghttp.RespondWith(200, bytes.Replace(infoBody, []byte("test-oauth-endpoint"), []byte("http://www.not-exist.com"), -1)),
+						),
+					)
 				})
 
-				It("returns an error", func() {
+				It("should have the endpoints populated and return error", func() {
+					Expect(cfc.GetEndpoints().AuthEndpoint).To(Equal("http://www.not-exist.com"))
```

Comment:
The testing of the Endpoints should be a separate positive test, not combined with some failure case.